### PR TITLE
Use Google Cloud AI Platform to train and serve the model

### PR DESCRIPTION
**Pull request by davidgasquez**
With this new package, we should be able to train, evaluate and serve a simple `Word2Vec` model.

### CODE REVIEW

File: model/requirements.txt [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+gensim
+google-api-python-client
+joblib
+numpy
+pandas
+pandas-gbq
+scikit-learn
+scikit-plot
+tensorflow
+tqdm
```
Comment (davidgasquez) [inline model/requirements.txt]: We should fix these dependencies once we get closer to the final model.

File: model/trainer/model.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        self.model = Word2Vec(
+            sentences=X,
+            size=self.size,
+            window=self.window,
+            min_count=self.min_count,
+            compute_loss=self.compute_loss,
+        )
+
+        self.loss = -self.model.get_latest_training_loss()
+        self.model = self.model.wv
```
Comment (davidgasquez) [inline model/trainer/model.py:25]: Overwriting the model with the word vectors to save memory and allow improvements like `nmap`.

File: model/trainer/task.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+from trainer.model import Word2VecEstimator
+
+METRIC_FILE_NAME = "eval_metrics.joblib"
+MODEL_FILE_NAME = "model.joblib"
+
+BASE_QUERY = """
+select
+  hashtags
+from `{table}`
+limit 100000
```
Comment (davidgasquez) [inline model/trainer/task.py]: I should remove this hardcoded value. :sweat_smile:

File: model/trainer/task.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    """
+
+    query = BASE_QUERY.format(table=full_table_path)
+    limit = " LIMIT {}".format(num_samples) if num_samples else ""
+    query += limit
+
+    data_df = pd.read_gbq(query, project_id=project_id, dialect="standard")
+    data_df.to_csv("data.csv", index=False)
+
+    # data_df = pd.read_csv("data.csv")
```
Comment (davidgasquez) [inline model/trainer/task.py]: Might be great to have a cache of the data locally. Will add some logic to download the dataset only once.

File: model/trainer/task.py [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+    """Upload local file to Google Cloud Storage.
+
+    Args:
+        local_path: (string) Local file
+        gcs_path: (string) Google Cloud Storage destination
+
+    Returns:
+        None
+    """
+    pass
```
Comment (davidgasquez) [inline model/trainer/task.py:76]: This should read datasets from GCS, so we can reduce the queries against QB.

File: model/trainer/task.py [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+        steps and model training
+        dataset: (pandas.DataFrame), DataFrame containing training data
+        output_dir: (string), directory that the trained model will be exported
+
+    Returns:
+        None
+    """
+    estimator.fit(dataset["hashtags"])
+
+    scores = model_selection.cross_val_score(estimator, dataset["hashtags"], cv=3)
```
Comment (davidgasquez) [inline model/trainer/task.py:108]: Right now, it only supports `loss` but we can create our own metric or reuse one.

**Review by davidgasquez — COMMENTED** [review]
Got the remote training working and the hyperparameter optimization too. Next step is serving the model!

File: model/requirements.txt [hunk 7]
```diff
@@ -0,0 +1,10 @@
+gensim==3.8.3
+google-api-python-client==1.7.8
+joblib==0.13.0
+numpy==1.16.5
+pandas==0.24.2
+pandas-gbq==0.13.2
+scikit-learn==0.20.4
+tensorflow==1.15.0
+tqdm==4.46.0
+cloudml-hypertune
```
Comment (davidgasquez) [inline model/requirements.txt:10]: Fixed the versions!

File: model/trainer/task.py [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+from trainer.model import Word2VecEstimator
+
+METRIC_FILE_NAME = "eval_metrics.joblib"
+MODEL_FILE_NAME = "model.joblib"
+
+BASE_QUERY = """
+select
+  hashtags
+from `{table}`
+limit 1000000
```
Comment (davidgasquez) [inline model/trainer/task.py:24]: We shouldn't be using BigQuery as the source. Once we get the final dataset, we could move it to GCS and use it from there.

File: model/trainer/task.py (hunk 6 above)
Comment (davidgasquez) [reply model/trainer/task.py:108]: Let's add mean average precision/recall at `k` once we merge this PR.

File: model/trainer/task.py [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+    """Testbed for running model training and evaluation."""
+    # Get data for training and evaluation
+
+    dataset = read_df_from_bigquery(flags.input, num_samples=flags.num_samples)
+
+    # Get model
+    estimator = Word2VecEstimator(flags)
+
+    # Create pipeline
+    pipeline = Pipeline([("word2vec", estimator)])
```
Comment (davidgasquez) [inline model/trainer/task.py:143]: This should live in the `model.py` file once we start adding more steps.

**Review by davidgasquez — COMMENTED** [review]
Heya @michael-erasmus! I think we can start merging this PR. It's only the most basic model but is working :sweat_smile:.  We can iterate from here!
